app/services/catalog/copy_portfolio_item.rb

```diff
@@ -0,0 +1,40 @@
+module Catalog
+  class CopyPortfolioItem
+    attr_reader :new_portfolio_item
+
+    def initialize(params)
+      @portfolio_item = PortfolioItem.find(params[:portfolio_item_id])
+      @to_portfolio = Portfolio.find(params[:portfolio_id] || @portfolio_item.portfolio_id)
+      portfolios_write_check
+    end
+
+    def process
+      @new_portfolio_item = make_copy
+      @to_portfolio.portfolio_items << @new_portfolio_item
+
+      self
+    end
+
+    private
+
+    def make_copy
+      @portfolio_item.dup.tap do |new_portfolio_item|
+        if new_portfolio_item.portfolio_id == @to_portfolio.id
+          new_portfolio_item.name = "Copy of " + new_portfolio_item.name
+        end
+
+        new_portfolio_item.save
+      end
+    end
+
+    def portfolios_write_check
```

Comment:
@lindgrenj6 Can this method be defined as a new method in the  RBAC mixin which you can call from here. The access check can happen at the class level and drill down further into the instance level. Here you are asking the question do I have write access to the portfolios class (collection).

Reply:
I agree that this check should not be handled in the service class.  I wonder though if the controller could be the caller? The `rbac_mixin` would do all required checks from the controller and not pass control to the Service Class until access was confirmed.  
Thoughts @mkanoor @lindgrenj6

Reply:
@syncrou Makes sense. This way it keeps all the RBAC checks outside of the service_object
Maybe we need to before_actions
- Can I read the Portfolio
- Can I create a new Portfolio

Reply:
Updated the `before_action` to do all the RBAC handling, that way it doesn't even make it into the action unless RBAC allows it.